### PR TITLE
Feature/process based guile session manager

### PR DESCRIPTION
**Pull request by Ophien**
Please make sure the below checklist is followed for Pull Requests.

- [x] Regression tests are added to tests/ for each new service
- [x] Explanatory MyServiceName.md is added to docs/ for each new service
- [x] README.md of the repository is updated to list each new service
- [x] General coding rules & commit guidelines as per our [CONTRIBUTING.md](https://github.com/singnet/wiki/tree/master/guidelines/CONTRIBUTING.md) document are followed

### CODE REVIEW

File: src/opencogServicesDef.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+#define OSS_OP_URL_LOADED 100
+#define OSS_OP_ATOMSPACE_RETRIEVED 101
+
+#define OSS_ERROR_CAN_NOT_DELETE_TEMP_FILE -100
+#define OSS_ERROR_CAN_NOT_LOAD_URL -101
+#define OSS_ERROR_CAN_NOT_LOAD_INVALID_SESSION -102
+#define OSS_ERROR_ATOMSPACE_CAN_NOT_BE_RETRIVED -104
+#define OSS_ERROR_FATAL_CAN_NOT_CREATE_PRIVATE_SESSION -500
+
```
Comment (andre-senna) [inline src/opencogServicesDef.h]: Delete everything below this line. Only #define's  should be kept here

File: src/opencogServicesDef.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+#include <string>
+#include <curl/curl.h>
+
+using namespace std;
+
+namespace opencogservices
+{
+extern void callSession(const string &rSessionProgram, const string &rReadPipe, const string &rWritePipe);
+extern int loadUrlFile(const string &rUrl, string &rOutputFileName);
```
Comment (andre-senna) [inline src/opencogServicesDef.h]: loadUrlFile() header should go in utils.h or something like that.

File: src/opencogServicesDef.h [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+extern size_t writeData(void *pPtr, size_t size, size_t nmemb, void *pStream);
```
Comment (andre-senna) [inline src/opencogServicesDef.h]: writeData() is used locally by loadUrlFile() so it should not be put in any header.

File: src/opencogServicesDef.h (hunk 2 above)
Comment (andre-senna) [reply src/opencogServicesDef.h]: Actually loadUrlFile() is only only in loadAtomeseFile() so what's the point of refactoring it?

File: src/opencogServicesDef.cc [hunk 4]
```diff
@@ -1,5 +1,9 @@
+
+#include <curl/curl.h>
 #include "opencogServicesDef.h"
 
+using namespace std;
+
 namespace opencogservices
```
Comment (andre-senna) [inline src/opencogServicesDef.cc]: remove namespace {} from .cc files and use "using namespace ..." instead

File: src/GuileSessionManager.cc [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+GuileSessionManager::GuileSessionManager(const char *guileSessionExePath)
+{
+	//TODO::change to hex generator
+	_idGen = GSM_INITIAL_ID;
+
+	// certify that there will be no running sessions from previously running mannagers
+	clearEnvironment();
+
+	// open a new sessions pid file
+	_pidFile = open(GSM_SESSION_OPENED_SESSIONS_FILE, O_CREAT | O_WRONLY);
```
[the remark this answers is not in the document]
Comment (Ophien) [reply src/GuileSessionManager.cc]: Replaced with ifstream to use getline().

File: src/OpencogSNETService.cc [hunk 6]
```diff
@@ -1,7 +1,56 @@
...
+	pCurlHandle = curl_easy_init();
+	curl_easy_setopt(pCurlHandle, CURLOPT_URL, rUrl.c_str());
+	curl_easy_setopt(pCurlHandle, CURLOPT_VERBOSE, 0L);
+	curl_easy_setopt(pCurlHandle, CURLOPT_NOPROGRESS, 1L);
+	curl_easy_setopt(pCurlHandle, CURLOPT_WRITEFUNCTION, writeData);
+	curl_easy_setopt(pCurlHandle, CURLOPT_WRITEDATA, pFile);
+
+	// if unable to handle url set error status, does not override file error
+	CURLcode errorCode = curl_easy_perform(pCurlHandle);
+	if (errorCode != CURLE_OK && status != UTILS_OP_ERROR_CAN_NOT_CREATE_FILE) {
```
Comment (andre-senna) [inline src/OpencogSNETService.cc]: You are missing the error message which should be passed back to the caller

File: src/GuileSessionManager.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (Ophien) [reply src/GuileSessionManager.cc]: Replaced by FILE* and fopen

File: src/Session.cc [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+{
+	return _writeFd;
+}
+
+int Session::getPid()
+{
+	return _pid;
+}
+
+string Session::getPidStr()
```
[the remark this answers is not in the document]
Comment (Ophien) [reply src/Session.cc]: Removed

**Pull request closed without merging**